M4 Report: Working doc (#189)
sopra-fs22-group-36/screw-your-neighbor-server: doc/m4/FS22-Group-36-M4-Report.md

there is only one highest rank in the trick.

When all the tricks in the match are played out, the players count the number of their tricks.

Points counting rules:\
If a player announced the number of tricks correctly, then he gets the number of announced tricks squared as 
positive points. Otherwise, the player gets the difference as negative points.

Then the cards are distributed for the next match. The dealer duty moves one player to the right, the number 
of cards that is distributed follows the sequence [5,4,3,2,1,2,3,4,5].

For the match with only 1 card there is a special rule. The players don’t see their own card but put their card 
with the front revealed on their forehead, so only the other players can see the card, but not the player himself. 
After the 9 matches, the player with the highest score wins.

## 2. Game view
The order of the screenshots corresponds to the flow of the game, if possible. Matches 4, 3, 2 and the last match 
are not listed. The only difference to the five-card match is the number of cards.
Depends on the result of the game, after the last match, only the winner can see a screw your neighbor cup, 
well-designed by Carmen. 

> ### X Technologies

Reviewer: Oh, right! Sorry for that.

#### X.x Spring Data REST

#### X.x Hibernate, JPA

#### X.x Jitsi API

#### X.x Mockito

#### X.x Git

> 

Reviewer: Did I forget technologies? Would you remove some of these ones?

### X.x High-level components
#### UI

#### Database

#### Game Logic (side effects)

#### Game Validations

#### Security

> 

Reviewer: Not sure, if I understood the term _component_ i.e. the task _Identify your project’s 3-5 main components._ in the right way. Maybe they mean with components also parts of the application that implement a specific logic, e.g. the stacking rule.

### 2.1 Enter the game
To enter the game, a player name must first be chosen. A new game can then be created in the lobby or an existing 
one can be joined. A player who creates a new game is automatically forwarded to the anteroom and can wait there 
until other players join the game or send an invitation link.

![Landing page](Landing_page.png)

![Lobby](Lobby.png)

![Room](Room.png)

### 2.2 Game playing
There are main three different types of game rounds. This is the standard round, where the highest card win the round. 
Afterwards is a stacked round where two cards have the same rank, and the next round will determine who has won both 
(or more) rounds. At the end there is the special round with only one card, where each player can't see his own card 
but all other ones. 

![Game announcing round](Announcing_stage.png)

![Game playing: standard round](Standard_round.png)

![Game playing: stacked round](Stacked_round.png)

![Game playing: one-card-round](Round_with_one_card.png)

### 2.2 Game ending
At the end of the game it is evaluated who has reached the most points. Different scenarios are visualized in 
different players.

> ![End of game: winner page](Winner_screen.png) [^1]

Reviewer: Fusszeile zerstört offenbar die Bildunterschrift. Im PDF hat dieses Bild als einziges keine Bildunterschrift, nur die Referenz zur Fusszeile.

> ![end of game: looser page](Loser_screen.png)

Reviewer: ```suggestion
![End of game: looser page](Loser_screen.png)
```

[^1]: Design by Carmen Kirchdorfer @CarmenUZH

## 3. Lessons learned
We have divided our learnings into three chapters, Challenges, success stories and notes on teamwork.

### 3.1 Challenges
- The most challenging part was to learn the tech stack in an appropriate time.
- The next challenge was to review the pull requests in a reasonable amount of time and then to implement the reviews.

> While this was a time-consuming process, it helps to understand the code and how the program works.

Reviewer: ```suggestion
While this was a time-consuming process, it helped to understand the code and how the program work.
```

Thus, with each pull request, several people knew how far the user stories were implemented.
- We have mixed feelings about using HATEOAS. On one hand every object had it's
[IRI](https://en.wikipedia.org/wiki/Internationalized_Resource_Identifier) as identifier, and the relations to other 

> objects was clear. On the other hand the serialization was not consistent with embedded entities, and we had

Reviewer: ```suggestion
objects were clear. On the other hand the serialization was not consistent with embedded entities, and we had
```

\centering

# SCREW YOUR NEIGHBOR

Report for Milestone 3 of Sopra FS22, Group 36

![cover (source B. Furrer)](../header/jass-image.png)

**Members**

Carmen Kirchdorfer (20-720-132)  
Salome Wildermuth (10-289-544)  
Beat Furrer, group leader (07-542-392)  
Lucius Bachmann (11-060-274)  
Moris Camporesi (19-764-349)

\newpage
\raggedright
## 1.  Introduction to the game
Our game can be played by two up to five people. The goal for every player is to gain as 
many points as possible during the game. Every player plays for himself, 
there are no teams. The game is played with a “Swiss Jass” card set of 36 cards with four suits.

> The game is divided into 9 matches (Runden) with changing number of distributed cards per player 

Reviewer: ```suggestion
The game is divided into 9 matches with changing number of distributed cards per player 
```

Reviewer: We use this term in the meantime mainly for the round per trick (not for the match anymore).

(5,4,3,2,1,2,3,4,5), and for every match points are distributed.

For the first match, every player gets 5 cards. 
Then one player after the other must announce how many tricks ("Stich", according to https://en.wikipedia.org/wiki/Jass) 
they will make in this match, 

> starting from the player right from the current dealer. 

Reviewer: ```suggestion
starting from the player to the right from the current dealer. 
```

> The last one announcing the number of his tricks for this match must announce several tricks such that the total 

Reviewer: ```suggestion
The last one announcing the number of his tricks for this match must announce as many tricks such that the total 
```

sum of announced tricks is not equal to the number of cards distributed per player in this match.

When the trick announcing round is completed, the game starts. The player to the right of the current dealer starts. 
The suits don’t matter in this game, just the rank, and there is no trump. The player who played the highest 

> rank in a match, i.e. wins the trick and has to deal the first card for the next trick.

Reviewer: ```suggestion
rank in a match, i.e. wins the trick, has to play the first card for the next trick.
```

the player who started the current trick has to start the next one too. If the last trick was stacked, the players 
who played the highest cards draw another card and continue the next trick with these. This may continue until 
there is only one highest rank in the trick.

When all the tricks in the match are played out, the players count the number of their tricks.

Points counting rules:\
If a player announced the number of tricks correctly, then he gets the number of announced tricks squared as 
positive points. Otherwise, the player gets the difference as negative points.

Then the cards are distributed for the next match. The dealer duty moves one player to the right, the number 
of cards that is distributed follows the sequence [5,4,3,2,1,2,3,4,5].

For the match with only 1 card there is a special rule. The players don’t see their own card but put their card 
with the front revealed on their forehead, so only the other players can see the card, but not the player himself. 
After the 9 matches, the player with the highest score wins.

## 2. Game view
The order of the screenshots corresponds to the flow of the game, if possible. Matches 4, 3, 2 and the last match 

> are not listed. The only difference to the five-card match is the number of cards.

Reviewer: What do you mean with _Matches 4, 3, 2, are not listed._ ?

Author: I mean rounds...thx Sherlok! ;-)

### 3.1 Challenges
- The most challenging part was learning in an appropriate time the technology stack. 
- The challenge was to review the pull requests in a reasonable amount of time and then to implement the reviews.
While this was a time-consuming process, it helps to understand the code and how the program works.
Thus, with each pull request, several people knew how far the user stories were implemented.
- Another challenge was the feedback on the milestones. Some reviews were difficult to understand. 

### 3.2 About teamwork
- What we enjoyed the most was playing the game in real life on a map and then later online every week, which has two 
huge advantages:
    1. It supported our team building process
    2. We test our game online every week

- Weekly meetings of the whole group at the Irchel have simplified communication. Meeting in person has the advantage 
that problems can be discussed bilaterally and the others can continue to work, but at the same time can listen with 
one ear.
	
- The use of Git and the clarification of how we work with this tool has made it much easier to edit the software 
efficiently, the main feature of Git we used was: 
    1. Pull-request

>     2. Feature-Branch 

Reviewer: ```suggestion
    2. Feature-Branch 
    3. Github Actions
```

> - As a small team building event, where of course the subject SoPra was extensively discussed, we met for the

Reviewer: haha :+1:

> "Sechseläuten". Coincidentally, there was a team member from the canton of Uri, which was also the guest canton.

Reviewer: And the team member is also a firefighter. He was really impressed and fascinated by the spectacle and the later Cervelat grill on the fire.

Reviewer: ```suggestion
"Sechseläuten". Coincidentally, there was a team member from the canton of Uri, which was also the guest canton.
![image](https://user-images.githubusercontent.com/45465496/169947777-4eb655b9-e3f0-4c64-8a7a-de4156249687.png)
```

### 4. Conclusion 
We are extremely satisfied with our performance as a group. We had an ideal mixture of very experienced software 
developers and those who acquired broad knowledge during the semester. In particular, the people for whom this 

> topic requires extensive experience in the field of programming. We all learned a lot about working together in a

Reviewer: Don't understand this sentence.

there is only one highest rank in the trick.

When all the tricks in the match are played out, the players count the number of their tricks.

Points counting rules:\
If a player announced the number of tricks correctly, then he gets the number of announced tricks squared as 
positive points. Otherwise, the player gets the difference as negative points.

Then the cards are distributed for the next match. The dealer duty moves one player to the right, the number 
of cards that is distributed follows the sequence [5,4,3,2,1,2,3,4,5].

For the match with only 1 card there is a special rule. The players don’t see their own card but put their card 
with the front revealed on their forehead, so only the other players can see the card, but not the player himself. 
After the 9 matches, the player with the highest score wins.

## 2. Game view
The order of the screenshots corresponds to the flow of the game, if possible. Matches 4, 3, 2 and the last match 
are not listed. The only difference to the five-card match is the number of cards.
Depends on the result of the game, after the last match, only the winner can see a screw your neighbor cup, 
well-designed by Carmen. 

> ## 3. Lessen learned

Reviewer: ```suggestion
## 3. Lessons learned
```
